addons/payment_payumoney/models/payumoney.py

```diff
@@ -0,0 +1,185 @@
+        assert acquirer.provider == 'payumoney'
+        def get_value(key):
+            if values.get(key):
+                return values[key]
+            return ''
+
+        if inout == 'in':
+            keys = "key|txnid|amount|productinfo|firstname|email|udf1|||||||||".split('|')
+            sign = ''.join('%s|' % (get_value(k)) for k in keys)
+            sign += acquirer.payu_salt
+        else:
+            keys = "|status||||||||||udf1|email|firstname|productinfo|amount|txnid".split('|')
+            sign = ''.join('%s|' % (get_value(k)) for k in keys)
+            sign = acquirer.payu_salt + sign + acquirer.merchant_id
+
+        shasign = hashlib.sha512(sign).hexdigest()
+        return shasign
+
+    @api.multi
+    def payumoney_form_generate_values(self, partner_values, tx_values):
+        self.ensure_one()
+        base_url = self.env['ir.config_parameter'].get_param('web.base.url')
+        payu_tx_values = dict(tx_values)
+        payu_tx_values.update({
+            'key': self.merchant_id,
+            'txnid': tx_values['reference'],
+            'amount': tx_values['amount'],
+            'productinfo': tx_values['reference'],
+            'firstname': partner_values['first_name'],
+            'email': partner_values['email'],
+            'phone': partner_values['phone'],
+            'service_provider': 'payu_paisa',
+            'surl': '%s' % urlparse.urljoin(base_url, PayuMoneyController._return_url),
+            'furl': '%s' % urlparse.urljoin(base_url, PayuMoneyController._exception_url),
+            'curl': '%s' % urlparse.urljoin(base_url, PayuMoneyController._cancel_url),
+        })
+
+        payu_tx_values['udf1'] = payu_tx_values.pop('return_url', '')
+        payu_tx_values['hash'] = self._payu_generate_sign(
+            self, 'in', payu_tx_values)
+        return partner_values, payu_tx_values
+
+    @api.multi
+    def payumoney_get_form_action_url(self):
+        self.ensure_one()
+        return self._get_payu_urls(self.environment)['payu_form_url']
+
+
+class PaymentTxPayuMoney(models.Model):
+    _inherit = 'payment.transaction'
+
+    payumoney_txnid = fields.Char(string='Transaction ID')
+    payumoney_id = fields.Char(string="Unique payment ID")
+
+# --------------------------------------------------
+# FORM RELATED METHODS
+# --------------------------------------------------
+
+    @api.model
+    def _payumoney_form_get_tx_from_data(self, data):
+        """ Given a data dict coming from payu, verify it and find the related
+        transaction record. """
+        reference, pay_id, shasign = data.get('txnid'), data.get(
+            'mihpayid'), data.get('hash')
+        if not reference or not pay_id or not shasign:
+            error_msg = 'Payu: received data with missing reference (%s) or pay_id (%s) or shashign (%s)' % (
+                reference, pay_id, shasign)
+            _logger.error(error_msg)
+            raise ValidationError(error_msg)
```

Comment:
Are you sure the validation error isn't going to be logged independenly by whatever ancestor layer catches it?